libs/chain-signatures/contract/src/lib.rs

```diff
@@ -62,93 +66,66 @@ impl Default for VersionedMpcContract {
     }
 }
 
-pub trait SignatureStateHandler {
-    fn add_request(&mut self, request: &SignatureRequest, data_id: CryptoHash);
-    fn get_pending_request(&mut self, request: &SignatureRequest) -> Option<YieldIndex>;
-    fn remove_request(&mut self, request: SignatureRequest) -> Result<(), Error>;
-}
-
-pub fn handle_signature_response<T: SignatureStateHandler>(
-    mpc_contract: &mut T,
-    request: &SignatureRequest,
-    response: &SignatureResponse,
-) -> Result<(), Error> {
-    if let Some(YieldIndex { data_id }) = mpc_contract.get_pending_request(request) {
-        env::promise_yield_resume(&data_id, &serde_json::to_vec(&response).unwrap());
-        Ok(())
-    } else {
-        Err(InvalidParameters::RequestNotFound.into())
-    }
-}
-
-pub fn update_contract_state_and_construct_signature_promise<T: SignatureStateHandler>(
-    mpc_contract: &mut T,
-    contract_signature_request: ContractSignatureRequest,
-) -> PromiseIndex {
-    let yield_promise = env::promise_yield_create(
-        "clear_state_on_finish",
-        &serde_json::to_vec(&(&contract_signature_request,)).unwrap(),
-        CLEAR_STATE_ON_FINISH_CALL_GAS,
-        GasWeight(0),
-        DATA_ID_REGISTER,
-    );
-
-    // Store the request in the contract's local state
-    let data_id: CryptoHash = env::read_register(DATA_ID_REGISTER)
-        .expect("read_register failed")
-        .try_into()
-        .expect("conversion to CryptoHash failed");
-
-    mpc_contract.add_request(&contract_signature_request.request, data_id);
-
-    // NOTE: there's another promise after the clear_state_on_finish to avoid any errors
-    // that would rollback the state.
-    env::promise_then(
-        yield_promise,
-        env::current_account_id(),
-        "return_signature_on_finish",
-        &[],
-        NearToken::from_near(0),
-        RETURN_SIGNATURE_ON_FINISH_CALL_GAS,
-    )
-}
-
 #[derive(BorshDeserialize, BorshSerialize, Debug)]
 pub struct MpcContractV1 {
     protocol_state: ProtocolContractState,
     pending_requests: LookupMap<SignatureRequest, YieldIndex>,
+    request_by_timestamp: Vector<(u64, SignatureRequest)>,
     proposed_updates: ProposedUpdates,
+    config: ConfigV1,
 }
 
-impl SignatureStateHandler for MpcContractV1 {
+impl MpcContractV1 {
+    fn remove_timed_out_requests(&mut self) -> u32 {
+        let max_to_remove = self.config.max_num_requests_to_remove;
+        let deadline = cmp::max(env::block_height(), self.config.request_timeout_blocks)
```

Comment:
`deadline` is fairly ambiguous, maybe something like `minimum_pending_request_height`?